src/python/pants/util/collections.py

```diff
@@ -9,3 +9,13 @@
 def combined_dict(*dicts):
   """Combine one or more dicts into a new, unified dict (dicts to the right take precedence)."""
   return {k: v for d in dicts for k, v in d.items()}
+
+
+def recursively_update(dict, dict2):
+  """dict.update but which merges child dicts (dict2 takes precedence where there's conflict)."""
+  for k, v in dict2.items():
+    if k in dict:
+      if isinstance(v, type(dict)):
```

Comment:
one quick thought here is that you're stomping on the `dict` builtin with your local variable naming here, which may be confusing if/when anyone needs to update this function in the future.

if you used e.g. `d` or `dict_` etc I think that'd be cleaner and you could avoid the `type(dict)` here in favor of direct use of `dict` as type.